pkg/openstackbackup/funcs.go

```diff
@@ -605,11 +607,94 @@ func CleanNamespace(
 	// OSP-D CRs can be mass-deleted
 	if len(crLists.OpenStackBaremetalSets.Items) > 0 {
 		foundRemaining = true
+		waitingOnBmhDeprovisionLabelAdded := false
+
+		// Special case: since OpenStackBaremetalSets might have associated BMHs currently provisioned,
+		// we need to check for those BMHs and wait until they are finished deprovisioning before continuing
+		// with deleting other resources.  If we don't, we can remove resources (such as networking CRs)
+		// that BMH deprovisioning currently depends upon and sabotage the deprovisioning by doing so (which
+		// will leave the BMHs stuck in the deprovisioning state).
+
+		for _, osBms := range crLists.OpenStackBaremetalSets.Items {
+			baremetalHostsList, err := ospdirectorv1beta1.GetBmhHosts(
+				ctx,
+				r.GetClient(),
+				"openshift-machine-api",
+				map[string]string{
+					common.OwnerControllerNameLabelSelector: shared.OpenStackBaremetalSetAppLabel,
+					common.OwnerNameLabelSelector:           osBms.Name,
+				},
+			)
+			if err != nil {
+				return false, err
+			}
+
+			// Annotate all the OpenStackBaremetalSet's BMHs with a special annotation to indicate to us
+			// that they must fully deprovision before we are done with them here
+			for _, bmh := range baremetalHostsList.Items {
+				labels := bmh.GetObjectMeta().GetLabels()
+				if _, ok := labels[shared.RequireBMHDeprovision]; !ok {
+					labels[shared.RequireBMHDeprovision] = ""
+					bmh.GetObjectMeta().SetLabels(labels)
+
+					waitingOnBmhDeprovisionLabelAdded = true
+
+					if err = r.GetClient().Update(ctx, &bmh); err != nil {
+						return false, err
+					}
+				}
+			}
+		}
+
+		// If we added our "RequireBMHDeprovision" label to any BMH, we need to return from the reconcile
+		// and then check back again later
+		if waitingOnBmhDeprovisionLabelAdded {
+			return false, nil
```

Comment:
do we need to return an requeueAfter since the backup controller is not watching the BMHs? I mean could there be a possibility that there is no reconcile event and we are just waiting for one to happen?

Reply:
cool